presto-main/src/main/java/com/facebook/presto/operator/JoinProbe.java

```diff
@@ -15,20 +15,124 @@
 
 import com.facebook.presto.spi.Page;
 import com.facebook.presto.spi.PageBuilder;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.type.Type;
 
-public interface JoinProbe
+import java.util.List;
+import java.util.Optional;
+import java.util.OptionalInt;
+
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+
+public class JoinProbe
```

Comment:
Can you take some time to see if this is the interface we want?  This interface was "forced" to make the code performant.  IIRC we had simple code before and I had to add this to make stuff fast, but also complex.

Reply:
Will remove `getOutputChannelCount` given it can be derived from `getOutputChannels`. Otherwise, seems there is little we can do unless we merge both the page builder and the probe into `LookupJoinOperator`.